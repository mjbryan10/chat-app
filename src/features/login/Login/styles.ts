import styled, { ThemeProps, Theme } from 'styled-components';

export const Container = styled.div`
   position: fixed;
   top: 0;
   right: 0;
   bottom: 0;
   left: 0;
   flex-flow: row nowrap;
   align-items: center;
   justify-content: center;
   display: flex;
   background: ${({ theme }: ThemeProps<Theme>) => theme.background.page ?? 'black'};
`;

export const Form = styled.form`
   padding: 1em;
   position: relative;
   flex-flow: column nowrap;
   display: flex;
   align-items: flex-end;
`;
