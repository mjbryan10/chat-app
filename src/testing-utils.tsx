import React from 'react';
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { dark } from './shared/theme';

const AllTheProviders = ({ children }: any) => {
  return (
    <ThemeProvider theme={dark}>
        {children}
    </ThemeProvider>
  )
  
}

const customRender = (ui?: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }