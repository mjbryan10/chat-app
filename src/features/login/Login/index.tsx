import React, { useState, FC, useEffect } from 'react';
import InputTextFieldWithLabel from '../../../components/InputTextFieldWithLabel';
import InputSubmitButton from '../../../components/InputSubmitButton';
import * as S from './styles';
import { useSelector, useDispatch } from 'react-redux';
import { selectLogin, loginUser, logoutUser } from '../loginSlice';
import Cookies from 'js-cookie';
import jwt from 'jsonwebtoken';
import Spinner from 'components/Spinner';

const secret = process.env.REACT_APP_JWT_SECRET;

const Login: FC = () => {
   const { isAuthenticated } = useSelector(selectLogin);
   const dispatch = useDispatch();

   const [status, setStatus] = useState<'loading' | 'ready'>('loading');
   const [name, setName] = useState('Wessel');
   const [id, setId] = useState(1);

   useEffect(() => {
      const cookieAuth = Cookies.get('auth_token');
      const resetLogin = () => {
         Cookies.remove('auth_token');
         dispatch(logoutUser());
      };
      if (cookieAuth && secret) {
         try {
            const decoded = jwt.verify(cookieAuth, secret) as
               | { [key: string]: any }
               | string;
            if (typeof decoded !== 'object') throw new Error('Token invalid.');
            if (decoded?.data?.id && decoded?.data?.name) {
               dispatch(loginUser({ id: decoded.data.id, name: decoded.data.name }));
            } else {
               throw new Error('Unable to decode token.');
            }
         } catch {
            resetLogin();
         }
      } else if (cookieAuth) {
         resetLogin();
      } else {
         setStatus('ready');
      }
   }, [dispatch]);

   const handleSubmit = (event: React.FormEvent) => {
      event.preventDefault();
      dispatch(loginUser({ id, name }));
      if (secret) {
         const token = jwt.sign({ data: { id, name } }, secret, { expiresIn: '7d' });
         Cookies.set('auth_token', token, { expires: 7 });
      }
   };

   return (
      <S.Container>
         {status === 'loading' ? (
            <Spinner />
         ) : (
            <S.Form action="submit" onSubmit={handleSubmit} data-testid="login-form">
               <InputTextFieldWithLabel
                  handleChange={(value: string) => setName(value)}
                  labelValue="Name"
                  initialValue="Wessel"
                  disabled
               />
               <InputTextFieldWithLabel
                  handleChange={(value: string) => setId(parseInt(value))}
                  labelValue="ID"
                  initialValue="1"
                  disabled
               />
               {isAuthenticated === false ? (
                  <S.ErorrText>
                     It appears we were unable to log you in with those details, please
                     try again later.
                  </S.ErorrText>
               ) : null}
               <InputSubmitButton value="Log in" />
            </S.Form>
         )}
      </S.Container>
   );
};

export default Login;
