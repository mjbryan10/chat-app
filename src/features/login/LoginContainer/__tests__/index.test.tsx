import React from 'react';
import Login from '..';
import { render, fireEvent } from 'testing-utils';
import axios from 'axios';
import jwt from 'jsonwebtoken';

jest.mock('jsonwebtoken', () => ({
   verify: jest.fn(() => {
      throw new Error();
   }),
   sign: jest.fn()
}));
jest.mock('axios');

describe('Login Component', () => {
   beforeEach(() => {
      (axios.get as jest.Mock).mockClear();
   });
   afterAll(() => jest.clearAllMocks());
   const setup = () => {
      const utils = render(<Login />);
      const nameInput = utils.getAllByTestId('input-text-field')[0];
      const idInput = utils.getAllByTestId('input-text-field')[1];
      const submitButton = utils.getByTestId('input-submit-button');
      const form = utils.getByTestId('login-form');
      return { nameInput, idInput, submitButton, form, ...utils };
   };

   test('should trigger axios request if form is submitted', () => {
      const { form } = setup();

      fireEvent.submit(form);

      expect(axios.get).toBeCalledTimes(1);
   });

   test('should trigger submit when `Log In` button is called', () => {
      const { submitButton } = setup();

      fireEvent.click(submitButton);

      expect(axios.get).toBeCalledTimes(1);
   });

   test('should render correctly', () => {
      const { asFragment } = render(<Login />);

      expect(asFragment()).toMatchSnapshot();
   });
});
