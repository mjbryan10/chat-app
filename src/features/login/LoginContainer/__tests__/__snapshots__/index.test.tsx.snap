// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm hYwnJW"
  >
    <div
      class="sc-AxirZ eigcKQ"
    >
      <form
        action="submit"
        class="sc-fzoLsD cVKEWn"
        data-testid="login-form"
      >
        <label
          class="sc-AxheI hKKxa-d"
        >
          Name
          <input
            class="sc-AxgMl ObLal"
            data-testid="input-text-field"
            placeholder=""
            value="Wessel"
          />
        </label>
        <label
          class="sc-AxheI hKKxa-d"
        >
          ID
          <input
            class="sc-AxgMl ObLal"
            data-testid="input-text-field"
            placeholder=""
            value="1"
          />
        </label>
        <input
          class="sc-AxmLO kpVHJF"
          data-testid="input-submit-button"
          type="submit"
          value="Log in"
        />
      </form>
      <button
        class="sc-AxiKw iEUgoF"
      >
        Change user
      </button>
    </div>
  </div>
</DocumentFragment>
`;
